### PR TITLE
1996 by Inlead: Implement the SPT project into DDBCMS.

### PR DESCRIPTION
**Pull request by inlead**
http://platform.dandigbib.org/issues/1996

### CODE REVIEW

File: modules/ding_mkws/ding_mkws.module [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+/**
+ * Implements hook_menu().
+ */
+function ding_mkws_menu() {
+  $items = array();
+
+  $items['admin/config/ding/mkws'] = array(
+    'title' => 'MKWS settings',
+    'description' => 'Settings and defaults for MKWS implementation',
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/ding_mkws.module]: I'd suggest you use a more descriptive title and description. Googling 'mkws' returns 'Milton Keynes Wargames Society' way ahead of 'MasterKey Widget Set'. And even THAT don't tell much about what it's actually used for...

File: modules/ding_mkws/ding_mkws.info [hunk 2]
```diff
@@ -0,0 +1,8 @@
+name = Ding MKWS
+description = Represents MasterKey Widget Set
+package = Ding!
+version = 7.x-1.0
+core = 7.x
+configure = admin/config/ding/moreinfo
+dependencies[] = libraries
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/ding_mkws.info]: there's also a dependency on ctools and page_manager

File: modules/ding_mkws/ding_mkws.admin.inc [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    form_set_error('pz2][proxy', t('Enter valid url for proxy.'));
+  }
+}
+
+/**
+ * Submit handler for saving pz2 settings.
+ */
+function ding_mkws_service_settings_form_submit($form, &$form_state) {
+  $pz2 = $form_state['values']['pz2'];
+  variable_set('ding_mkws_settings', $pz2);
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/ding_mkws.admin.inc]: I'd suggest adding a "your settings have been saved" user message of some sort

File: modules/ding_mkws/ding_mkws.admin.inc [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        }
+      }
+    }
+
+    $values['opac'][$key]['title'] = $items[$key]['title'];
+    $values['opac'][$key]['task_name'] = $items[$key]['task_name'];
+    $values['opac'][$key]['scope'] = $items[$key]['scope'];
+  }
+
+  variable_set('ding_mkws_global', $values);
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/ding_mkws.admin.inc]: I'd suggest adding a "your settings have been saved" user message of some sort

File: modules/ding_mkws/ding_mkws.admin.inc [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+          ),
+        );
+        field_create_instance($instance);
+      }
+    }
+    elseif ($instance) {
+      field_delete_instance($instance);
+    }
+  }
+  variable_set('ding_mkws_types', $values['types']);
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/ding_mkws.admin.inc]: I'd suggest adding a "your settings have been saved" user message of some sort

File: modules/ding_mkws/ding_mkws.fields.inc [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        }, $item);
+        $hash = md5(implode('', $imploded));
+        ding_mkws_include_all();
+        $markup = "<div data-hash='{$hash}'
+                          data-process='ProcessDataForNodeWidget'
+                          data-template='dingMkwsNodeWidget'
+                          class='ding-mkws-widget ding-mkws-node-widget'>
+                        </div>";
+        $element[0] = array(
+          '#markup' => $markup,
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/ding_mkws.fields.inc]: why hardcode the html when your're returning a render array?

File: modules/ding_mkws/ding_mkws.module [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    $hash = md5(implode('', $item));
+
+    ding_mkws_include_all();
+    $markup = "<div data-hash='{$hash}'
+            data-process='ProcessDataForNodeWidget' 
+            data-template='dingMkwsNodeWidget' 
+            class='ding-mkws-widget ding-mkws-node-widget'>
+        </div>";
+    $variables['text'] = array(
+      '#markup' => $markup,
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/ding_mkws.module]: why hardcode the html when your're returning a render array?

File: modules/ding_mkws/js/ding_mkws.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+ */
+
+/**
+ * Default config.
+ */
+var ding_mkws = {
+  active: false,
+  sort: 'retrieval',
+  settings: {},
+  spinner: '<div class="ispinner large gray animating">' +
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/js/ding_mkws.js]: why not use the standard DDB-CMS spinner?

File: modules/ding_mkws/plugins/content_types/ding_mkws_item_list.inc [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    if (!empty($global_settings['maxrecs'])) {
+      $params['maxrecs'] = $global_settings['maxrecs'];
+    }
+  }
+  else {
+    $params['term']['query'] = reset($panel_args);
+  }
+
+  $params['resources'] = (!empty($global_settings)) ? array_values($global_settings['resources']) : NULL;
+  $pane->content['#markup'] = "<div data-hash='{$hash}'
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/plugins/content_types/ding_mkws_item_list.inc]: why hardcode the html when you're returning a render array?

File: modules/ding_spt_statistics/ding_spt_statistics.module [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+  $items['spt/redirect'] = array(
+    'title' => 'Ding SPT redirect',
+    'description' => 'Count click and redirect to the external resource',
+    'type' => MENU_CALLBACK,
+    'access arguments' => array('access content'),
+    'page callback' => 'ding_spt_statistics_redirect',
+  );
+  $items['admin/reports/spt-statistics'] = array(
+    'title' => 'Ding SPT statistics',
```
Comment (JrgenGNielsen) [inline modules/ding_spt_statistics/ding_spt_statistics.module]: what does 'SPT' mean??

File: modules/ding_mkws/js/ding_mkws.theme.js [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+              "<div class='right-value'>"+
+                "<p class='ding-mkws-date'>" +
+                  '{{:date}}' +
+                "</p>" +
+              "</div>"+
+            "{{/if}}" +
+          "</div>" +
+        "</div>" +
+      "</div>" +
+      "<a class='ding-mkws-more-link' href='/search/meta/{{:more_link}}'>{{:more_text}}</a>" +
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/js/ding_mkws.theme.js]: href needs Drupal.settings.basePath + Drupal.settings.pathPrefix to work consistently

File: modules/ding_mkws/js/ding_mkws.theme.js [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+                  "<p class='ding-mkws-date'>" +
+                    '{{:date}}' +
+                  "</p>" +
+                "</div>"+
+              "{{/if}}" +
+            "</div>" +
+          "</div>" +
+        "{{/for}}</div>" +
+      "</div>" +
+      "<a href='/search/meta/{{:more_link}}'>{{:more_text}}</a>" +
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/js/ding_mkws.theme.js]: href needs Drupal.settings.basePath + Drupal.settings.pathPrefix to work consistently

File: modules/ding_mkws/plugins/content_types/ding_mkws_panel_widget.inc [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    'maxrecs' => (!empty($conf)) ? $conf['maxrecs'] : '',
+  );
+
+  $hash = md5(implode('', $params));
+
+  // Add resources after hash is generated
+  // in order to prevent array to string conversion notice.
+  $params['resources'] = (!empty($conf)) ? $conf['resources'] : array();
+
+  $pane->content['#markup'] = "<div data-hash='{$hash}'
```
Comment (JrgenGNielsen) [inline modules/ding_mkws/plugins/content_types/ding_mkws_panel_widget.inc]: why hardcode the html when you're returning a render array?

File: modules/mkdru_ding/js/mkdru_ding.theme.js [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+    html += '</div>' +
+      '</div>' +
+      '<div class="right-column left">';
+
+    if (hit["md-author"]) {
+      // expand on ; and reprint in the same form
+      var authors = hit["md-author"][0].split(';');
+      html += '<div class="creator"><span class="byline">' + Drupal.t('By') + ' </span>';
+      for (var i = 0; i < authors.length - 1; i++) {
+        html += '<a class="author" href="' + basePath + 'search/meta/' + specific_author_field + Drupal.theme.mkdruSafeTrim(authors[i]) + '">' + authors[i] + '</a> ;';
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/js/mkdru_ding.theme.js]: href needs Drupal.settings.basePath + Drupal.settings.pathPrefix to work consistently

File: modules/mkdru_ding/js/mkdru_ding.theme.js [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+      }
+      html += '<a class="author" href="' + basePath + 'search/meta/' + specific_author_field + Drupal.theme.mkdruSafeTrim(authors[authors.length - 1]) + '">' + authors[authors.length - 1] + '</a>';
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/js/mkdru_ding.theme.js]: ditto

File: modules/mkdru_ding/js/mkdru_ding.theme.js [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+    } else if (hit["md-journal-title"]) {
+      html += '<div class="mkdru-result-journal">' + hit["md-journal-title"];
+      html += '</div><p/>';
+    }
+    if (dhit["md-subject"] && dhit["md-subject"].length > 0) {
+      var search_query;
+      html += '<div class="mkdru-result-subject"><p>';
+      for (var i = 0; i < dhit["md-subject"].length - 1; i++) {
+        search_query = 'su=&quot;' + specific_subject_field + dhit["md-subject"][i] + '&quot;';
+        html += '<a href="' + basePath + 'search/meta/' + search_query + '">' + dhit["md-subject"][i] + '</a> ';
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/js/mkdru_ding.theme.js]: href needs Drupal.settings.basePath + Drupal.settings.pathPrefix to work consistently

File: modules/mkdru_ding/js/mkdru_ding.theme.js [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+      }
+      search_query = 'su=&quot;' + specific_subject_field + dhit["md-subject"][dhit["md-subject"].length - 1] + '&quot;';
+      html += '<a href="' + basePath + 'search/meta/' + search_query + '">' + dhit["md-subject"][dhit["md-subject"].length - 1] + '</a></p></div>';
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/js/mkdru_ding.theme.js]: href needs Drupal.settings.basePath + Drupal.settings.pathPrefix to work consistently

File: modules/mkdru_ding/plugins/content_types/search_facets.inc [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+ */
+function mkdru_ding_search_facets_render($subtype, $conf, $args, $context) {
+  $block = new stdClass();
+  $block->title = t('Facets');
+
+  $mkdru_config = variable_get('mkdru_defaults');
+  $facets = empty($mkdru_config['facets']) ? array() : $mkdru_config['facets'];
+  $block->content = '';
+  foreach ($facets as $facet => $facet_options) {
+    $block->content .= theme('mkdru_block_facet', array('class' => 'mkdru-facet-' . $facet, 'name' => $facet_options['displayName']));
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/plugins/content_types/search_facets.inc]: calling the theme() function directly is strongly discouraged. 
It is preferable to replace direct calls to the theme() function with calls to drupal_render() by passing a render array with a #theme key to drupal_render(), which in turn calls theme().
(see: https://api.drupal.org/api/drupal/includes!theme.inc/function/theme/7.x)

File: modules/mkdru_ding/mkdru_ding.module [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+function mkdru_ding_search_execute($keys = '', $conditions = NULL) {
+  $state = array();
+
+  if ($keys) {
+    $state['query'] = $keys;
+  }
+
+  $mkdru_defaults = variable_get('mkdru_defaults', array());
+  $path = drupal_get_path('module', 'mkdru_ding');
+  theme('mkdru_js', array(
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/mkdru_ding.module]: the output from theme() isn't used... does this actually do anything? also, calling the theme() function directly is strongly discouraged.

File: modules/mkdru_ding/plugins/content_types/search_results.inc [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+  // Build the content type block.
+  $block = new stdClass();
+
+  $results = '';
+  $conditions = array();
+
+  // Collect the search results.
+  $results = search_data($keys, $info['module'], $conditions);
+  $block->content = theme('mkdru_results');
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/plugins/content_types/search_results.inc]: calling the theme() function directly is strongly discouraged.

File: modules/mkdru_ding/mkdru_ding.module [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+    'path' => 'meta',
+    'conditions_callback' => 'mkdru_ding_search_conditions_callback',
+  );
+}
+
+/**
+ * Implements hook_search_page().
+ */
+function mkdru_ding_search_page($results) {
+  $output['prefix']['#markup'] = theme('mkdru_results');
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/mkdru_ding.module]: calling the theme() function directly is strongly discouraged.

File: modules/mkdru_ding/mkdru_ding.module [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+/**
+ * Implements hook_block_view().
+ */
+function mkdru_ding_block_view($delta = '') {
+  $block = array();
+
+  switch ($delta) {
+    case 'mkdru_ding_search_results':
+      $block['subject'] = t('mkdru ding search results');
+      $block['content'] = '<div class="mkdru-pager"></div><div class="mkdru-counts"></div><p></p><div class="mkdru-result-list"></div><div class="mkdru-pager"></div><div class="mkdru-counts"></div>';
```
Comment (JrgenGNielsen) [inline modules/mkdru_ding/mkdru_ding.module]: html shouldn't be hardcoded - instead, put it in a template file, and make a theme function for it.

**Comment by JrgenGNielsen** [comment]
Please note that a "drush feature revert all" will remove Ding MKWS item list from the search-ting_search panel page. 
And that, as far as I can tell, "Authenticate on server" needs to be checked (sp_server_auth = TRUE) - it defaults to NULL

File: modules/ding_mkws/tests/MkwsConfigTest.php [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    $this->setBrowser($this->config->getBrowser());
+    $this->setBrowserUrl($this->config->getUrl());
+    sleep(5);
+  }
+
+  public function testMkwsModules() {
+    $this->open("/");
+    $this->click("//div[@id='page']/header/section/div/ul/li[2]/a/span");
+    $this->type("id=edit-name", "admin");
+    $this->type("id=edit-pass", "1234");
```
Comment (guddo) [inline modules/ding_mkws/tests/MkwsConfigTest.php]: This is not valid.

File: modules/ding_mkws/tests/SPTTestConfig.class.inc [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+ */
+class SPTTestConfig {
+  private $browser;
+  private $url;
+  private $user;
+  private $pass;
+
+  public function __construct() {
+    $this->browser = isset($_SERVER['SPTTEST_BROWSER']) ? $_SERVER['SPTTEST_BROWSER'] : '*firefox';
+    $this->url = isset($_SERVER['SPTTEST_URL']) ? $_SERVER['SPTTEST_URL'] : 'http://spt-v2.ddbcms.dev.inlead.dk/';
```
Comment (guddo) [inline modules/ding_mkws/tests/SPTTestConfig.class.inc]: You should not commit these variables.

File: modules/ding_spt_statistics/tests/SPTTestConfig.class.inc [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+ */
+class SPTTestConfig {
+  private $browser;
+  private $url;
+  private $user;
+  private $pass;
+
+  public function __construct() {
+    $this->browser = isset($_SERVER['SPTTEST_BROWSER']) ? $_SERVER['SPTTEST_BROWSER'] : '*firefox';
+    $this->url = isset($_SERVER['SPTTEST_URL']) ? $_SERVER['SPTTEST_URL'] : 'http://spt-v2.ddbcms.dev.inlead.dk/';
```
Comment (guddo) [inline modules/ding_spt_statistics/tests/SPTTestConfig.class.inc]: You should not commit these variables.

File: modules/ding_spt_statistics/tests/SptModConfigTest.php [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+    $this->setBrowser($this->config->getBrowser());
+    $this->setBrowserUrl($this->config->getUrl());
+    sleep(5);
+  }
+
+  public function testSptModules() {
+    $this->open("/");
+    $this->click("//div[@id='page']/header/section/div/ul/li[2]/a/span");
+    $this->type("id=edit-name", "admin");
+    $this->type("id=edit-pass", "1234");
```
Comment (guddo) [inline modules/ding_spt_statistics/tests/SptModConfigTest.php]: This is not valid.

File: modules/ding_spt_statistics/tests/SptStatisticTest.php [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+    $is_present_elem = $this->isElementPresent('id=ding-spt-statistics-date-picker-form');
+    // Verify if are the table statistics.
+    sleep(2);
+    if ($is_present_elem) {
+      $this->assertTrue($this->isElementPresent('css=div.form-item-date-from'));
+      $this->assertTrue($this->isElementPresent('css=div.form-item-date-to'));   
+      $this->assertTrue($this->isElementPresent('css=input#edit-reset'));   
+      sleep(5);
+    }
+    else echo 'The tabell statistic is not!!!';
```
Comment (guddo) [inline modules/ding_spt_statistics/tests/SptStatisticTest.php]: What does this mean?

File: modules/mkdru_ding/tests/SPTTestConfig.class.inc [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+ */
+class SPTTestConfig {
+  private $browser;
+  private $url;
+  private $user;
+  private $pass;
+
+  public function __construct() {
+    $this->browser = isset($_SERVER['SPTTEST_BROWSER']) ? $_SERVER['SPTTEST_BROWSER'] : '*firefox';
+    $this->url = isset($_SERVER['SPTTEST_URL']) ? $_SERVER['SPTTEST_URL'] : 'http://spt-v2.ddbcms.dev.inlead.dk/';
```
Comment (guddo) [inline modules/mkdru_ding/tests/SPTTestConfig.class.inc]: You should not commit these variables.

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 4bcf97a3640df8098de68441deb498d9da46e999
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 4bcf97a3640df8098de68441deb498d9da46e999
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by JrgenGNielsen** [comment]
reviewed and accepted

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: ec770c7480df4bdc531bbd32242a028d4a223e09
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: ec770c7480df4bdc531bbd32242a028d4a223e09
  * The subject should be kept around 50 characters long. The subject must not be more than 69 characters long. Strive for about 50 characters.

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by inlead** [comment]
Moved to https://github.com/ding2/ding2/pull/650.

**Pull request closed without merging**